src/cpp/session/modules/rmarkdown/SessionRmdNotebook.cpp

```diff
@@ -0,0 +1,670 @@
+/*
+ * SessionRmdNotebook.cpp
+ *
+ * Copyright (C) 2009-16 by RStudio, Inc.
+ *
+ * Unless you have received this program directly from RStudio pursuant
+ * to the terms of a commercial license agreement with RStudio, then
+ * this program is licensed to you under the terms of version 3 of the
+ * GNU Affero General Public License. This program is distributed WITHOUT
+ * ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
+ * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
+ * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
+ *
+ */
+
+#include "SessionRmdNotebook.hpp"
+
+#include <iostream>
+
+#include <boost/foreach.hpp>
+
+#include <r/RJson.hpp>
+#include <r/RExec.hpp>
+
+#include <core/Exec.hpp>
+#include <core/FileSerializer.hpp>
+#include <core/Algorithm.hpp>
+#include <core/json/Json.hpp>
+#include <core/json/JsonRpc.hpp>
+#include <core/text/CsvParser.hpp>
+#include <core/StringUtils.hpp>
+
+#include <session/SessionModuleContext.hpp>
+#include <session/SessionSourceDatabase.hpp>
+#include <session/SessionOptions.hpp>
+
+#define kChunkDefs        "chunk_definitions"
+#define kChunkDocId       "doc_id"
+#define kChunkId          "chunk_id"
+#define kChunkLibDir      "lib"
+#define kChunkOutputPath  "chunk_output"
+#define kChunkUrl         "url"
+#define kChunkConsole     "console"
+
+#define kChunkConsoleInput  0
+#define kChunkConsoleOutput 1
+#define kChunkConsoleError  3
+
+#define kChunkTypeNone    0
+#define kChunkTypeOutput  1
+#define kChunkTypeConsole 2
+
+using namespace rstudio::core;
+
+namespace rstudio {
+namespace session {
+namespace modules {
+namespace rmarkdown {
+namespace notebook {
+
+namespace {
+
+// the ID of the doc / chunk currently executing console commands (if any)
+std::string s_consoleChunkId, s_consoleDocId, s_activeConsole;
+
+// whether we're currently connected to console events
+bool s_consoleConnected = false;
+
+// A notebook .Rmd is accompanied by a sidecar .Rnb.cached folder, which has
+// the following structure:
+//
+// - foo.Rmd
+// + foo.Rnd.cached
+//   - chunks.json
+//   - cwiaiw9i4f0.html
+//   + cwiaiw9i4f0_files
+//     - plot.png
+//   - c0aj9vhk0cz.html
+//   - cjz0958jgzh.csv
+//   + lib
+//     + htmlwidgets
+//       - htmlwidget.js
+// 
+// That is:
+// - each chunk has an ID and is represented by one of the following:
+//   - an HTML file (.html) with accompanying dependencies indicating
+//     chunk output, or 
+//   - a CSV file (.csv) indicating console output
+// - the special file "chunks.json" indicates the location of the chunks
+//   in the source .Rmd
+// - the special folder "lib" is used for shared libraries (e.g. scripts upon
+//   which several htmlwidget chunks depend)
+
+
+FilePath chunkCacheFolder(const std::string& docPath, const std::string& docId)
```

Comment:
I think we should store the chunk cache side by side with the document itself (so that e.g. you can delete a chunk cache by deleting the parent directory). This implies the following:

1. Create the directory side by side and be sure it's hidden (you can call core::system::makeFileHidden for this)

2. Use userSettings().contextId with the folder to make sure that each user/machine context gets a separate cache.